java/card-bot/src/main/java/com/google/hangouts/chat/card/Bot.java

```diff
@@ -0,0 +1,209 @@
+ */
+package com.google.hangouts.chat.card;
+
+// [START card-bot]
+import com.fasterxml.jackson.databind.JsonNode;
+import com.google.api.services.chat.v1.model.*;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.ResponseBody;
+import org.springframework.web.bind.annotation.RestController;
+
+import java.util.*;
+import java.util.logging.Logger;
+import java.util.stream.Stream;
+
+@SpringBootApplication
+@RestController
+public class Bot {
+    private static final Logger logger = Logger.getLogger(Bot.class.getName());
+
+    private static final String INTERACTIVE_TEXT_BUTTON_ACTION = "doTextButtonAction";
+    private static final String INTERACTIVE_IMAGE_BUTTON_ACTION = "doImageButtonAction";
+    private static final String INTERACTIVE_BUTTON_KEY = "originalMessage";
+    private static final String HEADER_IMAGE = "https://goo.gl/5obRKj";
+    private static final String BOT_NAME = "Card Bot Java";
+    private static final String REDIRECT_URL = "https://goo.gl/kwhSNz";
+
+    public static void main(String[] args) {
+        SpringApplication.run(Bot.class, args);
+    }
+
+    /**
+     * Handles a GET request to the /bot endpoint.
+     *
+     * @param event Event from chat.
+     * @return Message
+     */
+    @PostMapping("/")
+    @ResponseBody
+    public Message onEvent(@RequestBody JsonNode event) {
+        Message reply = new Message();
+        String eventType = event.get("type").asText();
+        switch (eventType) {
+            case "ADDED_TO_SPACE":
+                JsonNode space = event.get("space");
+                String spaceType = space.get("type").asText();
+                if ("ROOM".equals(spaceType)) {
+                    String displayName = space.get("displayName").asText();
+                    String replyText = String.format("Thanks for adding me to %s", displayName);
+                    reply.setText(replyText);
+                } else {
+                    String displayName = event.get("user").get("displayName").asText();
+                    String replyText = String.format("Thanks for adding me to a DM, %s!", displayName);
+                    reply.setText(replyText);
+                }
+                break;
+            case "MESSAGE":
+                Card card = createCardResponse(event.get("message").get("text").asText());
+                reply.setCards(Collections.singletonList(card));
+                break;
+            case "CARD_CLICKED":
+                // Get the custom action name and custom parameter value out of the event object.
+                String actionName = event.get("action").get("actionMethodName").asText();
+                String customParameterValue = event.get("action").get("parameters").get(0).get("value").asText();
```

Comment:
What is a customParameterValue? Can we rename this userClickType or something less generic?